src/blockdata/witness.rs

```diff
@@ -0,0 +1,209 @@
+use consensus::{Decodable, Encodable, WriteExt, serialize};
+use VarInt;
+use io::{self, Read, Write};
+use consensus::encode::Error;
+use prelude::*;
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]  //TODO manually implement?
+pub struct Witness {
```

Comment:
Does it make sense to have a "pointers to the start of each witness element" vec as well? I guess given most witnesses are like 2-5 elements long maybe its not worth it, but would be cool to add a benchmark for iterating a lightning HTLC transaction witness and just test.

Reply:
Pointers to the start of each witness element would allow efficient index access but require another Vec.
I think for witness is not worth it because, as you said usually the witness is a few elements.

Reply:
Maybe have offsets in [`SmallVec`](https://docs.rs/smallvec/1.7.0/smallvec/)?
We could also optimistically assume that transactions are smaller than 4M bytes, so we could cram offsets into `struct U24(u8, u8, u8);`

Reply:
> Maybe have offsets in SmallVec?
https://github.com/rust-bitcoin/rust-bitcoin/pull/672#discussion_r721120421 
Also, having offsets would allow efficient indexed access, but mutable access it's another story... In the end I don't think there are really use case requiring that. I think I would not add offsets here even with tinyvec/smallvec.

> We could also optimistically assume that transactions are smaller than 4M bytes, so we could cram offsets into struct U24(u8, u8, u8);

the overall weight of the struct would not change because of padding (on most common 64 bits arch), I would leave this optimization in case we need other fields in the struct